core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -229,9 +229,16 @@ abstract class AbstractFetcherThread(name: String,
     }
   }
 
-  protected def truncateOnFetchResponse(epochEndOffsets: Map[TopicPartition, EpochEndOffset]): Unit = {
+  // Visibility for unit tests
+  protected[server] def truncateOnFetchResponse(epochEndOffsets: Map[TopicPartition, EpochEndOffset]): Unit = {
     inLock(partitionMapLock) {
-      val ResultWithPartitions(fetchOffsets, partitionsWithError) = maybeTruncateToEpochEndOffsets(epochEndOffsets, Map.empty)
+      // Partitions may have been removed from the fetcher while the thread was waiting for fetch
+      // response. Filter out removed partitions while holding `partitionMapLock` to ensure that we
+      // don't update state for any partition that may have already been migrated to another thread.
+      val filteredEpochEndOffsets = epochEndOffsets.filter { case (tp, _) =>
+        partitionStates.contains(tp)
+      }
```

Comment:
Did you consider pushing this down into `maybeTruncateToEpochEndOffsets`? This would avoid creating an extra collection here.

Reply:
@dajac Thanks for the review, updated.